File: frontend/src/lib/api/roster.ts
import { api } from '$lib/api';

export interface RosterRole {
	id: number;
	name: string;
	description: string | null;
	person_count: number;
	created_at: string;
	updated_at: string;
}

export interface RosterRoleCreate {
	name: string;
	description?: string;
}

export interface RosterRoleUpdate {
	name?: string;
	description?: string;
}

export interface PersonRosterRole {
	id: number;
	person_id: number;
	role_id: number;
	created_at: string;
}

export interface RosterTemplateSettings {
	keep_assignee: boolean;
	auto_open_hours: number;
	reminder_hours: number[];
	allow_self_assign: boolean;
}

export interface RosterTemplateSlotCreate {
	role_id: number;
	label: string;
	sort_order?: number;
	min_persons?: number;
	max_persons?: number;
}

export interface RosterTemplateSlot {
	id: number;
	role_id: number;
	role_name?: string;
	label: string;
	sort_order: number;
	min_persons: number;
	max_persons: number;
	created_at: string;
	updated_at: string;
}

export interface RosterTemplate {
	id: number;
	name: string;
	description?: string;
	ministry_id?: number;
	mass_time_id?: number;
	event_id?: number;
	recurrence_rule: string;
	recurrence_end?: string;
	settings: RosterTemplateSettings;
	is_active: boolean;
	slots: RosterTemplateSlot[];
	slot_count: number;
	created_at: string;
	updated_at: string;
}

export interface RosterTemplateCreate {
	name: string;
	description?: string;
	ministry_id?: number;
	recurrence_rule?: string;
	settings?: RosterTemplateSettings;
	is_active?: boolean;
	slots?: RosterTemplateSlotCreate[];
}

export interface RosterAssignment {
	id: number;
	instance_id: number;
	slot_id: number;
	person_id: number;
	person_name?: string;
	slot_label?: string;
	role_name?: string;
	status: string;
	assigned_by?: number;
	assigned_at?: string;
	accepted_at?: string;
	declined_at?: string;
	completed_at?: string;
	cancelled_at?: string;
	notes?: string;
	created_at: string;
}

export interface RosterAssignmentCreate {
	instance_id: number;
	slot_id: number;
	person_id: number;
}

export interface RosterInstance {
	id: number;
	template_id: number;
	template_name?: string;
	date: string;
	status: string;
	generated_at?: string;
	published_at?: string;
	completed_at?: string;
	assignments: RosterAssignment[];
	created_at: string;
}

function qs(params: Record<string, string | number | boolean | undefined>): string {
	const pairs = Object.entries(params)
		.filter(([, v]) => v !== undefined)
		.map(([k, v]) => `${encodeURIComponent(k)}=${encodeURIComponent(String(v))}`);
	return pairs.length ? `?${pairs.join('&')}` : '';
}

export const rosterApi = {
	// Roles
	listRoles: () => api.get<RosterRole[]>('/roster/roles'),
	createRole: (data: RosterRoleCreate) => api.post<RosterRole>('/roster/roles', data),
	updateRole: (id: number, data: RosterRoleUpdate) => api.put<RosterRole>(`/roster/roles/${id}`, data),
	deleteRole: (id: number) => api.delete(`/roster/roles/${id}`),
	assignRole: (roleId: number, personId: number) =>
		api.post<PersonRosterRole>(`/roster/roles/${roleId}/assign`, { person_id: personId, role_id: roleId }),
	removeRole: (roleId: number, personId: number) =>
		api.delete(`/roster/roles/${roleId}/persons/${personId}`),

	// Templates
	listTemplates: (params?: { ministry_id?: number; is_active?: boolean }) =>
		api.get<RosterTemplate[]>(`/roster/templates${qs({ ministry_id: params?.ministry_id, is_active: params?.is_active })}`),
	getTemplate: (id: number) => api.get<RosterTemplate>(`/roster/templates/${id}`),
	createTemplate: (data: RosterTemplateCreate) => api.post<RosterTemplate>('/roster/templates', data),
	updateTemplate: (id: number, data: Partial<RosterTemplateCreate>) =>
		api.put<RosterTemplate>(`/roster/templates/${id}`, data),
	deleteTemplate: (id: number) => api.delete(`/roster/templates/${id}`),
	duplicateTemplate: (id: number) => api.post<RosterTemplate>(`/roster/templates/${id}/duplicate`),

	// Instances
	generateInstance: (templateId: number, date: string) =>
		api.post<RosterInstance>(`/roster/templates/${templateId}/generate${qs({ date })}`),
	listInstances: (params?: { date_from?: string; date_to?: string; ministry_id?: number }) =>
		api.get<RosterInstance[]>(`/roster/instances${qs(params || {})}`),
	getInstance: (id: number) => api.get<RosterInstance>(`/roster/instances/${id}`),
	publishInstance: (id: number) => api.put<RosterInstance>(`/roster/instances/${id}/publish`, {}),
	cancelInstance: (id: number) => api.put<RosterInstance>(`/roster/instances/${id}/cancel`, {}),
	completeInstance: (id: number) => api.put<RosterInstance>(`/roster/instances/${id}/complete`, {}),

	// Assignments (admin)
	assignPerson: (instanceId: number, data: RosterAssignmentCreate) =>
		api.post<RosterAssignment>(`/roster/instances/${instanceId}/assign`, data),
	removeAssignment: (id: number) => api.delete(`/roster/assignments/${id}`),

	// Parish aggregate
	getParishAggregate: (date: string) => api.get(`/roster/parish${qs({ date })}`),
};

File: backend/app/schemas/roster.py
"""Pydantic schemas for the Roster System."""

from datetime import date, datetime
from typing import Annotated, Optional

from pydantic import BaseModel, ConfigDict, Field


# ---------------------------------------------------------------------------
# RosterRole
# ---------------------------------------------------------------------------
class RosterRoleBase(BaseModel):
    name: Annotated[str, Field(max_length=100)]
    description: Annotated[Optional[str], Field(max_length=2000)] = None


class RosterRoleCreate(RosterRoleBase):
    pass


class RosterRoleUpdate(BaseModel):
    name: Annotated[Optional[str], Field(max_length=100)] = None
    description: Annotated[Optional[str], Field(max_length=2000)] = None


class RosterRoleResponse(RosterRoleBase):
    model_config = ConfigDict(from_attributes=True)
    id: int
    person_count: int = 0
    created_at: datetime
    updated_at: datetime


# ---------------------------------------------------------------------------
# PersonRosterRole
# ---------------------------------------------------------------------------
class PersonRosterRoleCreate(BaseModel):
    person_id: int
    role_id: int


class PersonRosterRoleResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    person_id: int
    role_id: int
    created_at: datetime


# ---------------------------------------------------------------------------
# RosterTemplate
# ---------------------------------------------------------------------------
class RosterTemplateSettings(BaseModel):
    """Validated settings JSONB shape."""

    keep_assignee: bool = False
    auto_open_hours: int = Field(default=168, ge=1, le=720)  # 1 week default, max 30 days
    reminder_hours: list[int] = [48, 24]
    allow_self_assign: bool = True


class RosterTemplateBase(BaseModel):
    name: Annotated[str, Field(max_length=200)]
    description: Annotated[Optional[str], Field(max_length=5000)] = None
    ministry_id: Optional[int] = None
    mass_time_id: Optional[int] = None
    event_id: Optional[int] = None
    recurrence_rule: str = "none"  # none, weekly, biweekly, monthly
    recurrence_end: Optional[date] = None
    settings: RosterTemplateSettings = Field(default_factory=RosterTemplateSettings)
    is_active: bool = True


class RosterTemplateCreate(RosterTemplateBase):
    slots: list["RosterTemplateSlotCreate"] = []


class RosterTemplateUpdate(BaseModel):
    name: Annotated[Optional[str], Field(max_length=200)] = None
    description: Annotated[Optional[str], Field(max_length=5000)] = None
    ministry_id: Optional[int] = None
    mass_time_id: Optional[int] = None
    event_id: Optional[int] = None
    recurrence_rule: Optional[str] = None
    recurrence_end: Optional[date] = None
    settings: Optional[RosterTemplateSettings] = None
    is_active: Optional[bool] = None


class RosterTemplateResponse(RosterTemplateBase):
    model_config = ConfigDict(from_attributes=True)
    id: int
    slots: list["RosterTemplateSlotResponse"] = []
    slot_count: int = 0
    created_at: datetime
    updated_at: datetime


# ---------------------------------------------------------------------------
# RosterTemplateSlot
# ---------------------------------------------------------------------------
class RosterTemplateSlotBase(BaseModel):
    role_id: int
    label: Annotated[str, Field(max_length=200)]
    sort_order: int = 0
    min_persons: int = 1
    max_persons: int = 1


class RosterTemplateSlotCreate(RosterTemplateSlotBase):
    pass


class RosterTemplateSlotUpdate(BaseModel):
    role_id: Optional[int] = None
    label: Annotated[Optional[str], Field(max_length=200)] = None
    sort_order: Optional[int] = None
    min_persons: Optional[int] = None
    max_persons: Optional[int] = None


class RosterTemplateSlotResponse(RosterTemplateSlotBase):
    model_config = ConfigDict(from_attributes=True)
    id: int
    role_name: Optional[str] = None  # populated from relationship
    created_at: datetime
    updated_at: datetime


# ---------------------------------------------------------------------------
# RosterInstance
# ---------------------------------------------------------------------------
class RosterInstanceCreate(BaseModel):
    """Request schema for creating a roster instance from a template."""
    template_id: int
    date: date
    publish: bool = False  # If true, publish immediately
    notes: Optional[str] = None


class RosterInstanceResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    template_id: int
    template_name: Optional[str] = None
    date: date
    status: str
    generated_at: Optional[datetime] = None
    published_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    assignments: list["RosterAssignmentResponse"] = []
    created_at: datetime


class RosterInstanceStatusUpdate(BaseModel):
    status: str  # draft, published, completed, cancelled


# ---------------------------------------------------------------------------
# RosterAssignment
# ---------------------------------------------------------------------------
class RosterAssignmentCreate(BaseModel):
    instance_id: int
    slot_id: int
    person_id: int


class RosterAssignmentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    instance_id: int
    slot_id: int
    person_id: int
    person_name: Optional[str] = None  # populated from relationship
    slot_label: Optional[str] = None
    role_name: Optional[str] = None
    status: str
    assigned_by: Optional[int] = None
    assigned_at: Optional[datetime] = None
    accepted_at: Optional[datetime] = None
    declined_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    cancelled_at: Optional[datetime] = None
    notes: Optional[str] = None
    created_at: datetime


class RosterAssignmentStatusUpdate(BaseModel):
    status: str  # accepted, declined, completed, cancelled
    notes: Optional[str] = None


# ---------------------------------------------------------------------------
# RosterSwapRequest
# ---------------------------------------------------------------------------
class RosterSwapCreate(BaseModel):
    assignment_id: int
    to_person_id: int
    notes: Optional[str] = None


class RosterSwapResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    assignment_id: int
    from_person_id: int
    from_person_name: Optional[str] = None
    to_person_id: int
    to_person_name: Optional[str] = None
    status: str
    requested_at: datetime
    resolved_at: Optional[datetime] = None
    notes: Optional[str] = None
